### PR TITLE
Multiple tests

### PR DESCRIPTION
**Pull request by zacheliason**
I think it might be easiest to go over this on our next zoom call but here is the pull request for multiple tests!

I have a couple questions as well, especially regarding this paragraph in the trello card you wrote for me.
> In many cases, the instructor does not specify test code. We just execute their code and see if it matches the expected output. I'm not sure how to handle this. But here's an idea...We can allow the instructor to specify zero tests. If they do that, their solution must produce non-empty output. If it produces empty output, we throw an error when they try to save the entire exercise. If it produces output, we add an entry to the tests table with empty test code and show the expected output. We probably should display this a little differently on exercise.html (and student_exercise.html) than a regular test. In fact, we should probably do this all the time, even if some tests have been specified. Perhaps I am overcomplicating this...

### CODE REVIEW

**Comment by srp33** [comment]
Okay, let's discuss on Monday

File: back_ends/r/exec.R [hunk 1]
```diff
@@ -28,24 +28,44 @@ if (file.exists(check_code_file_path)) {
...
+
+setwd(file.path(tests_dir_path, "outputs"))
+outputs_dir <- getwd()
+
+tests <- list.files(path=tests_dir_path, pattern="test*", full.names=TRUE, recursive=FALSE)
+for (i in seq_along(tests)) {
+    setwd(file.path(outputs_dir, paste("test_", i, sep="")))
+    test_code <- readChar(tests[i], file.info(tests[i])$size)
+
+    # need to find a way to suppress printing to stdout while saving test code output so test outputs don't show up in submission output
```
Comment (srp33) [inline back_ends/r/exec.R:61]: I may be missing it, but do we need to check whether the tests dir exists as we are for the other back ends?

File: front_end/helper.py [hunk 2]
```diff
@@ -110,15 +110,10 @@ def exec_code(settings_dict, code, exercise_basics, exercise_details, request=No
...
     response_dict = json.loads(response.content)
-    return response_dict["text_output"], response_dict["image_output"]
-
-def check_exercise_output(exercise_details, actual_text, actual_image):
-    if exercise_details["back_end"] == "any_response" and (len(actual_text) > 0 or len(actual_image) > 0):
-        return "", True
-
-    if exercise_details["output_type"] == "txt":
-        expected_text = exercise_details["expected_text_output"]
+    return response_dict["text_output"], response_dict["image_output"], json.loads(response_dict["tests"])
```
Comment (srp33) [inline front_end/helper.py:113]: Shouldn't there be a different text_output and image_output for each test? What does response_dict["tests"] contain? I assume this is different from the tests in exercise_details? Maybe rename the response key here so that it is more intuitive what is included in it and make a comment that clarifies it.

File: front_end/html/edit_exercise.html [hunk 3]
```diff
@@ -226,44 +231,31 @@ <h2>Create exercise</h2>
         </div>
 
         <div class="shadow-box">
+            <div id="test_upload">
+                <p><strong>Tests:</strong></p>
+                <p>Test code can be used to run one or more tests using the student's code. It helps test the integrity of their code rather than its ability to produce the expected output.</p>
```
Comment (srp33) [inline front_end/html/edit_exercise.html:236]: Please change to "You can write one or more tests that evaluate the integrity of the student's code. Oftentimes, tests are used when the instructor has asked students to write a function, and each test invokes the function using certain parameters. Either way, the test code will be invoked *after* the student's code. For each test, you may indicate whether the test code should be visible to students. Hidden tests help to ensure that the student cannot guess the expected output  Generally, we recommend including at least one visible test and one hidden test. We recommend that hidden tests use similar approaches as visible tests."

Comment (srp33) [reply front_end/html/edit_exercise.html:236]: Also, I forgot whether we are adding the ability for students to receive partial credit if they pass some tests but not others?

File: middle_layer/main.py [hunk 4]
```diff
@@ -80,14 +88,45 @@ def exec(info: ExecInfo):
...
+                            try:
+                                # python script saves here for some reason
+                                with open(f"{tmp_dir_path}/tests/test_image_output_{i}", "rb") as output_file:
+                                    test_image_output = encode_image_bytes(output_file.read())
+                            except:
+                                # image failed to save
+                                test_image_output = ""
+                tests.append({"test": i, "text_output": test_text_output, "image_output": test_image_output})
+
         if info.output_type == "jpg":
```
Comment (srp33) [inline middle_layer/main.py:117]: Please add a comment that describes what this part does.

**Review by srp33 — CHANGES_REQUESTED** [review]

File: front_end/html/edit_exercise.html (hunk 3 above)
Comment (zacheliason) [reply front_end/html/edit_exercise.html:236]: We haven't added that, but I could if you'd like!

**Comment by zacheliason** [comment]
> Okay, let's discuss on Monday

That would be great, thanks!

File: front_end/html/edit_exercise.html (hunk 3 above)
Comment (srp33) [reply front_end/html/edit_exercise.html]: Please remind me that we should discuss this as well in our next meeting.

File: back_ends/r/exec.R [hunk 5]
```diff
@@ -28,24 +28,46 @@ if (file.exists(check_code_file_path)) {
...
+    tests <- list.files(path=tests_dir_path, pattern="test*", full.names=TRUE, recursive=FALSE)
+    for (i in seq_along(tests)) {
+        setwd(file.path(outputs_dir, paste("test_", i, sep="")))
+        test_code <- readChar(tests[i], file.info(tests[i])$size)
+
+        # need to find a way to suppress printing to stdout while saving test code output so test outputs don't show up in submission output
+        # Need to allow for multiline output!!!! FIXME
+        if (output_type == "txt") {
+          filename <- "text_output"
+          out <- invisible(suppressMessages(suppressWarnings(suppressPackageStartupMessages(eval(parse(text=test_code))))))
```
Comment (zacheliason) [inline back_ends/r/exec.R:65]: So I found a pretty big bug in my code here, previously we'd been printing the ```eval(parse(text=code))``` line that runs the actual code, and I assumed I could just assign the code output to a variable called ```out.``` unfortunately, this only catches the last line of output. I'm not as familiar with R so I was hoping there's some way to run the code and then save it to a variable which can be written to a file? 

Some alternatives that come to mind should that prove impossible would be not differentiating R exercise outputs by test (ie just printing the output of the solution + all tests to stdout), or going into the back end and reconfiguring things so that it hits the back end once for each test. Neither of those sound particularly optimal to me so I was hoping you would know of a better way to do things in R! Thanks!

File: back_ends/python/exec.py [hunk 6]
```diff
@@ -5,28 +5,56 @@
...
+my_plt_saver.close()""")
+
+                result = subprocess.run(f"python {test_code_path}", shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT).stdout.decode()
+                result = re.sub(r"Traceback \(most recent call last\)", r"Traceback (most recent call last)", result)
+
+                # save test text output under outputs in case of image traceback
+                with open(filename, "w") as test_output:
+                    test_output.write(result)
 
 if output_type == "jpg":
```
[the remark this answers is not in the document]
Comment (zacheliason) [reply back_ends/python/exec.py:50]: I'm confused, this code is writing output to a file, not creating a directory. Should I be more explicit about that in my wording?

File: front_end/html/edit_exercise.html (hunk 3 above)
Comment (zacheliason) [reply front_end/html/edit_exercise.html]: I'm afraid I forgot to remind you, but I will on Thursday!

File: back_ends/python/exec.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (srp33) [reply back_ends/python/exec.py:50]: I am not sure. Perhaps I was thinking about the line that has `if os.path.isdir(tests_dir_path):`. But if you could add some comments in general, that would help.

File: back_ends/r/exec.R (hunk 5 above)
Comment (srp33) [reply back_ends/r/exec.R:65]: We may need to change the logic to be more like what we are doing in `exec.py` (python back end). We are not reading the code into a variable and then executing it. We are leaving the code in a file and using the `python` command to execute it. We should be able to do something different using the `Rscript` command. The only gotcha is that with the current code in `exec.r`, we are using the "suppress" functions to remove unnecessary output. Instead of the "suppress" functions, we could use an approach like this: https://stackoverflow.com/questions/16194212/how-to-suppress-warnings-globally-in-an-r-script.

Do you think this approach would work?

**Review by srp33 — CHANGES_REQUESTED** [review]

File: blank_image.txt [hunk 7]
```diff
@@ -0,0 +1 @@
...KACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Z
```
Comment (zacheliason) [inline blank_image.txt:1]: Not sure if this is a huge deal or whether this scenario would ever even come up, but when you write a function that generates images to be tested with test cases, the solution 'expected_image_output' will save a blank white image. This white image is for comparison so that only non blank images are saved to the database.

File: front_end/content.py [hunk 8]
```diff
@@ -1524,13 +1569,43 @@ def save_exercise(self, exercise_basics, exercise_details):
...
             exercise_basics["exists"] = True
 
+            sql = '''DELETE FROM tests
+                     WHERE course_id = ?
+                       AND assignment_id = ?
+                       AND exercise_id = ?'''
+
+            self.execute(sql, [exercise_basics["assignment"]["course"]["id"], exercise_basics["assignment"]["id"], exercise_basics["id"]])
+
+            if exercise_details["tests"] != []:
```
Comment (srp33) [inline front_end/content.py:1602]: Can we just check the length rather than comparing against an empty list? Might not make a difference but might be faster.

File: front_end/content.py [hunk 9]
```diff
@@ -1539,18 +1614,41 @@ def save_presubmission(self, course, assignment, exercise, user, code):
...
+            if partner_id:
+                submission_id = self.get_next_submission_id(course, assignment, exercise, partner_id)
+                self.execute(sql, [int(course), int(assignment), int(exercise), partner_id, int(submission_id), code, text_output, image_output, passed, datetime.now(), user, tests])
+
+                if len(tests) > 0:
+                    for test in tests:
+                        self.execute(test_sql, [int(course), int(assignment), int(exercise), user, int(submission_id), test["text_output"], test["image_output"]])
+
+        except:
+            print("HERE")
```
Comment (srp33) [inline front_end/content.py:1650]: Can this be removed?

File: front_end/helper.py [hunk 10]
```diff
@@ -110,15 +110,12 @@ def exec_code(settings_dict, code, exercise_basics, exercise_details, request=No
...
-    return response_dict["text_output"], response_dict["image_output"]
-
-def check_exercise_output(exercise_details, actual_text, actual_image):
-    if exercise_details["back_end"] == "any_response" and (len(actual_text) > 0 or len(actual_image) > 0):
-        return "", True
-
-    if exercise_details["output_type"] == "txt":
-        expected_text = exercise_details["expected_text_output"]
+    # 'text_output' and 'image_output' refer to the output produced by the code written by the student, while 'tests' is a dict containing image and text outputs specific to each test case written by the instructor
+    # tests must be converted from json to a list of dicts
```
Comment (srp33) [inline front_end/helper.py:114]: This is super nitpicky, but please use correct capitalization and punctuation in comments.

File: front_end/helper.py [hunk 11]
```diff
@@ -110,15 +110,12 @@ def exec_code(settings_dict, code, exercise_basics, exercise_details, request=No
...
+    return response_dict["text_output"], response_dict["image_output"], json.loads(response_dict["tests"])
 
+def compare_outcome(expected_text, actual_text, expected_image, actual_image, output_type):
```
Comment (srp33) [inline front_end/helper.py:117]: Please use a more descriptive function name. Maybe `compare_outputs`.

File: front_end/helper.py [hunk 12]
```diff
@@ -131,25 +128,37 @@ def check_exercise_output(exercise_details, actual_text, actual_image):
...
-        #return diff_output, diff_percent == 0 # Pass if they are identical.
+
+def check_exercise_output(exercise_details, actual_text, actual_image, tests):
+    if exercise_details["back_end"] == "any_response" and (len(actual_text) > 0 or len(actual_image) > 0):
+        return "", True, []
+
+    test_outcomes = []
+    diff_output, passed = compare_outcome(exercise_details["expected_text_output"], actual_text, exercise_details["expected_image_output"], actual_image, exercise_details["output_type"])
+    expected_test_outputs = exercise_details["tests"]
+    for i in range(len(expected_test_outputs)):
```
Comment (srp33) [inline front_end/helper.py:157]: Please add some vertical whitespace.

File: front_end/html/edit_exercise.html [hunk 13]
```diff
@@ -226,44 +235,31 @@ <h2>Create exercise</h2>
         </div>
 
         <div class="shadow-box">
+            <div id="test_upload">
+                <p><strong>Tests:</strong></p>
+                <p>You can write one or more tests that evaluate the integrity of the student's code. Oftentimes, tests are used when the instructor has asked students to write a function, and each test invokes the function using certain parameters. Either way, the test code will be invoked after the student's code. For each test, you may indicate whether the test code should be visible to students. Hidden tests help to ensure that the student cannot guess the expected output Generally, we recommend including at least one visible test and one hidden test. We recommend that hidden tests use similar approaches as visible tests.</p>
```
Comment (srp33) [inline front_end/html/edit_exercise.html:240]: Please add a period before "Generally"

File: front_end/html/exercise.html [hunk 14]
```diff
@@ -350,9 +350,15 @@ <h6>Instructions:</h6>
             {% end %}
         </div>
 
-        {% if exercise_details["show_test_code"] and len(exercise_details["test_code"]) > 0 %}
+        {% if exercise_details["tests"] != [] %}
```
Comment (srp33) [inline front_end/html/exercise.html:353]: Check length rather than compare against empty list?

File: front_end/html/exercise.html [hunk 15]
```diff
@@ -468,9 +474,28 @@ <h6 style="display: inline">Expected output:</h6>
...
-                {% else %}
+                {% elif exercise_details["expected_image_output"] != "" %}
                     <p><figure class="img"><img src="data:image/jpg;base64,{{ exercise_details['expected_image_output'] }}" width="100%"/></figure></p>
+                {% else %}
+                    <p><pre>[Expected solution does not produce an image.]</pre></p>
                 {% end %}
+
+                <div id="expected_tests">
+                {% if len(exercise_details["tests"]) > 0 %}
+                {% for i in range(len(exercise_details["tests"])) %}
```
Comment (srp33) [inline front_end/html/exercise.html:485]: Add some vertical whitespace.

Should the `for` loop be indented within the `if` statement?

File: front_end/html/unavailable_assignment.html [hunk 16]
```diff
@@ -9,6 +9,8 @@ <h2>{{ assignment_basics["title"] }}</h2>
             <p>This assignment closed on {{ due_date }}.</p>
         {% elif error == "restricted_ip" %}
             <p>This assignment can only be completed from computers that have an allowed IP address. The IP address for this computer does not match any of the allowed addresses; please consult the instructor for additional instructions.</p>
+        {% elif error == "insufficient_access" %}
+            <p>insufficient access to view this page. Try logging in from an administrator account.</p>
```
Comment (srp33) [inline front_end/html/unavailable_assignment.html:13]: Please change to, "You do not have permission to access this page."

File: middle_layer/main.py [hunk 17]
```diff
@@ -51,27 +51,37 @@ def exec(info: ExecInfo):
...
-                test_file.write(info.test_code)
+        # save each test case in its own file under the directory 'tests'
+        if len(info.tests) > 0:
+            for i in range(len(info.tests)):
+                filename = f"{tmp_dir_path}/tests/test_{int(i + 1)}"
+                # preemptively create directory for test outputs
+                outputname = f"{tmp_dir_path}/tests/outputs/test_{int(i + 1)}/image_output" if info.output_type == "jpg" else f"{tmp_dir_path}/tests/outputs/test_{int(i + 1)}/text_output"
+                os.makedirs(os.path.dirname(filename), exist_ok=True)
+                os.makedirs(os.path.dirname(outputname), exist_ok=True)
+                with open(filename, "w") as test_file:
```
Comment (srp33) [inline middle_layer/main.py:62]: Please add some vertical whitespace.

File: middle_layer/main.py [hunk 18]
```diff
@@ -80,14 +90,50 @@ def exec(info: ExecInfo):
...
+                    for output_path in outputs:
+                        if "image" not in output_path:
+                            # read text output
+                            with open(f"{tmp_dir_path}/tests/outputs/{test_output}/{output_path}") as output_file:
+                                test_text_output = output_file.read()
+                        else:
+                            # read text output in case it contains traceback from an error writing the image
+                            with open(f"{tmp_dir_path}/tests/outputs/{test_output}/{output_path}") as output_file:
+                                test_text_output = output_file.read()
+                            try:
```
Comment (srp33) [inline middle_layer/main.py:112]: Please add some vertical whitespace.

**Review by srp33 — CHANGES_REQUESTED** [review]

File: back_ends/r/exec.R [hunk 19]
```diff
@@ -28,24 +30,45 @@ if (file.exists(check_code_file_path)) {
...
+    ggsave(paste("/sandbox/test_image_output_", i, sep=""), dpi = 150, device = "jpeg")
+  }
 }
 
 if (output_type == "txt") {
-  suppressMessages(suppressWarnings(suppressPackageStartupMessages(eval(parse(text=code)))))
+  system(paste("Rscript", code_file_path)) # run code using Rscript
 } else {
+  code <- readChar(code_file_path, file.info(code_file_path)$size)
   suppressMessages(suppressWarnings(suppressPackageStartupMessages(exec_jpg(code))))
```
[the remark this answers is not in the document]
Comment (zacheliason) [reply back_ends/r/exec.R:51]: I found out that all the suppressMessages and Warnings wrappers around ```exec_jpg``` are actually necessary to avoid getting notifications that say ```Saving 7 x 7 in image
`geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'``` so I left them in.

**Comment by zacheliason** [comment]
I finished all the changes requested for multiple_tests!

**Review by srp33 — APPROVED** [review]
